app-info/pwa-metainfo-generator.py

```diff
@@ -199,28 +150,50 @@ def _get_app_id(self):
         hashed_url = hashlib.sha1(self.url.encode("utf-8")).hexdigest()
         return "org.gnome.Software.WebApp_" + hashed_url
 
+    def _add_comment(self, element, text):
+        text = textwrap.fill(text, break_long_words=False, break_on_hyphens=False)
+        text = f' {text} '.replace('\n', '\n       ')
+        comment = ET.Comment(text)
+        element.append(comment)
+
     def _add_name(self, app_component):
-        name = self.config.get('name')
-        if not name:
-            # No name was configured. First try the OpenGraph name.
-            name = og_property_from_head(self.soup, "site_name")
-            if not name:
-                # No OpenGraph name. Try the manifest name, allowing the property to be
-                # chosen.
-                name_property = self.config.get('name_property')
-                if name_property is None:
-                    # Short name seems more suitable in practice
-                    if "short_name" in self.manifest:
-                        name_property = "short_name"
-                    else:
-                        name_property = "name"
-                elif name_property not in ('name', 'short_name'):
-                    raise ValueError(
-                        "name_property must be set to 'name' or 'short_name', not '{}'"
-                        .format(name_property)
-                    )
-                name = self.manifest[name_property]
-        ET.SubElement(app_component, 'name').text = name
+        self._add_comment(
+            app_component,
+            textwrap.dedent("""\
+            A human-readable name for this web app. See
+            https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-name
+            for details.
+            """),
+        )
+
+        # Build a list of found names.
+        names = []
+        og_name = og_property_from_head(self.soup, "site_name")
+        if og_name:
+            names.append((og_name, "OpenGraph name"))
+        manifest_name = self.manifest.get("name")
+        if manifest_name:
+            names.append((manifest_name, "Manifest name"))
+        manifest_short_name = self.manifest.get("short_name")
+        if manifest_short_name:
+            names.append((manifest_short_name, "Manifest short_name"))
+
+        num_names = len(names)
+        if num_names > 1:
+            self._add_comment(
+                app_component,
+                "Multiple names detected. Delete all but one name element below.",
```

Comment:
What if there are multiple names, but they’re identical? Is that case likely enough to make it worthwhile to make the generator silently coalesce identical names?

Same for some of the other elements below.

Reply:
I suppose it would be helpful to not output duplicate elements. However, since one of my goals here is to show all the scraped data, I think I'd want to update the comments to indicate if the value came from multiple sources. I think I have an idea for a cleanish way to do that.

Reply:
I reworked it a bit so that the values get collected into a dictionary with a list of comments as values, which deduplicates the values. I tried it out on a couple of the GNOME PWA sites and it did help.